### PR TITLE
Add mariadb

### PR DESCRIPTION
**Pull request by ZStriker19**
# Integration
This PR adds support for [`mariadb`](https://github.com/DataDog/documentation/pull/10838).

## Links

- Integration docs:
  - https://mariadb.com/docs/clients/mariadb-connectors/connector-python/
  - https://mariadb-corporation.github.io/mariadb-connector-python/usage.html
- Corp docs PR: https://github.com/DataDog/documentation/pull/10838

Added the PR to make changes to the test docker container. This will need to be merged before mariadb tests will pass in CircleCI: https://github.com/DataDog/dd-trace-py/pull/2584

## Checklist

- [x] Usage and configuration documentation added in `__init__.py`, `docs/index.rst` and `docs/integrations.rst`.
- [x] [Corp docs](https://github.com/Datadog/documentation) PR to add new integration to documentation.
- [x] Span metadata
  - [x] Service (use [`int_service`](https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/contrib/trace_utils.py#L55) or [`ext_service`](https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/contrib/trace_utils.py#L87)).
  - [x] Span type should be one of [these](https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/ext/__init__.py#L7).
  - [x] Resource
  - [x] Measured tag
  - [x] Sample analytics
- [x] Global configuration
  - [x] `ddtrace.config` entry is specified.
  - [x] Environment variables are provided for config options.
- [x] Instance configuration
  - [x] Pin overriding.
  - [x] Service name override (if applicable).
- [x] Tests
  - [x] Use `pytest` fixtures found in `tests/conftest.py` or the test helpers on `TracerTestCase` if
    writing `unittest` style test cases.
  - [x] Tests are provided for all the above.
  - [x] Tests are added to CI (`.circleci/config.yml`).
  - [x] Functionality is maintained from original library.
  - [x] Patch test cases are added (see `test_django_patch.py` for an example).
  - [x] All Python versions that the library supports are tested.
  - [x] All significant library versions (including the latest) are tested. This typically includes every minor release going back a few years.

### CODE REVIEW

File: ddtrace/__init__.py
Comment (brettlangdon) [inline ddtrace/__init__.py:12]: can you revert this change, seems unnecessary

File: ddtrace/contrib/mariadb/__init__.py
Comment (brettlangdon) [inline ddtrace/contrib/mariadb/__init__.py:2]:
```suggestion
The mariadb integration instruments the mariadb library to trace mariadb queries.
```

File: ddtrace/contrib/mariadb/patch.py
Comment (brettlangdon) [inline ddtrace/contrib/mariadb/patch.py:50]: missing new line at end of the file

File: ddtrace/contrib/mariadb/patch.py
Comment (brettlangdon) [inline ddtrace/contrib/mariadb/patch.py:26]: we want to add a guard here to ensure we don't double patch mariadb.

https://github.com/DataDog/dd-trace-py/blob/0dee2fcc7ecbdde7bb95b015b66b638660e349ef/ddtrace/contrib/redis/patch.py#L26-L28

We set the attribute on the module instead of just checking if it is already an ObjectProxy because if some other vendor has wrapped it it might already be an ObjectProxy, in those cases we still want to wrap it.

File: ddtrace/contrib/mariadb/patch.py
Comment (brettlangdon) [inline ddtrace/contrib/mariadb/patch.py:30]: We want to use a similar guard here as well.

https://github.com/DataDog/dd-trace-py/blob/0dee2fcc7ecbdde7bb95b015b66b638660e349ef/ddtrace/contrib/redis/patch.py#L47-L48

File: ddtrace/contrib/mariadb/patch.py
Comment (brettlangdon) [inline ddtrace/contrib/mariadb/patch.py:43]: This formatting looks off, have you run this file through `black`?

`riot run -s fmt`

File: docker-compose.yml
Comment (brettlangdon) [inline docker-compose.yml:158]: missing newline at the end of the file.

File: riotfile.py
Comment (brettlangdon) [inline riotfile.py:927]: we probably want to support `0.9.x` as well since it isn't that old (2020)

Comment (brettlangdon) [inline riotfile.py:927]:
```suggestion
                            "~=0.9",  # Most recent `0.x.x` release
                            "~=1.0.0",  # Most recent `1.0.x` release
                            "~=1.0",  # Most recent `1.x.x` release
```

File: riotfile.py
Comment (brettlangdon) [inline riotfile.py:921]:
```suggestion
            command="pytest {cmdargs} tests/contrib/mariadb",
```

File: riotfile.py
Comment (brettlangdon) [reply riotfile.py:927]: I now see that `0.9.x` is listed as beta 🤔 what kind of effort is it to support `0.9.x` if it is super easy/basically the same support, then maybe it is worth including... otherwise we can add support later if requested

File: tests/contrib/mariadb/test_backwards_compatibility.py
Comment (brettlangdon) [inline tests/contrib/mariadb/test_backwards_compatibility.py:8]: should we just get rid of this file?

File: tests/contrib/mariadb/test_mariadb.py
Comment (brettlangdon) [inline tests/contrib/mariadb/test_mariadb.py:21]: make this to run `black` on this file.

File: tests/contrib/mariadb/test_mariadb.py
Comment (brettlangdon) [inline tests/contrib/mariadb/test_mariadb.py:26]: this should probably be:

```python
try:
    yield tracer
finally:
  unpatch()
```

we want to be sure that no matter what we unpatch after every test case

File: tests/contrib/mariadb/test_mariadb.py
Comment (brettlangdon) [inline tests/contrib/mariadb/test_mariadb.py:35]: should we use the config you added for connection info?

File: tests/contrib/mariadb/test_mariadb.py
Comment (brettlangdon) [inline tests/contrib/mariadb/test_mariadb.py:45]: does the connection support context management?

```python
with get_connection(tracer) as connection:
    yield connection
```

File: tests/contrib/mysql/test_mysql.py
Comment (brettlangdon) [inline tests/contrib/mysql/test_mysql.py:62]: why this change, seems unrelated? this shouldn't be changing

File: tests/snapshots/tests.contrib.mariadb.test_mariadb.test_simple_query_snapshot.snap
Comment (brettlangdon) [inline tests/snapshots/tests.contrib.mariadb.test_mariadb.test_simple_query_snapshot.snap:1]: there should be more of these for the other snapshot tests you added right?

File: tests/contrib/mariadb/test_mariadb.py
Comment (brettlangdon) [inline tests/contrib/mariadb/test_mariadb.py:253]: I think mostly all of these can probably be snapshot tests

File: tests/contrib/mariadb/test_mariadb.py
Comment (brettlangdon) [inline tests/contrib/mariadb/test_mariadb.py:311]: we have another fixture you can use so you don't need to build a class/TracerTestCase

https://github.com/DataDog/dd-trace-py/blob/0dee2fcc7ecbdde7bb95b015b66b638660e349ef/tests/conftest.py#L22-L39

File: ddtrace/contrib/mariadb/patch.py
Comment (P403n1x87) [inline ddtrace/contrib/mariadb/patch.py:38]: We have the helper `get_argument_value` for this in `ddtrace.utils`

File: tests/contrib/mysql/test_mysql.py
Comment (P403n1x87) [reply tests/contrib/mysql/test_mysql.py:62]: `trace_fetch_method` has been removed from `dbapi2` and integration-specific configuration should be used. Is this causing issues with tests?

File: tests/contrib/mariadb/test_backwards_compatibility.py
Comment (ZStriker19) [reply tests/contrib/mariadb/test_backwards_compatibility.py:8]: Yes!

File: riotfile.py
Comment (ZStriker19) [reply riotfile.py:927]: Tried this but it looks like there are setup issues. Going to not include for now.:  

ERROR: Command errored out with exit status 2:
     command: /root/project/.riot/venv_py387/bin/python -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-6un8vj37/mariadb_43af4698120740069f3c51294b346354/setup.py'"'"'; __file__='"'"'/tmp/pip-install-6un8vj37/mariadb_43af4698120740069f3c51294b346354/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-s3b3ojzd
         cwd: /tmp/pip-install-6un8vj37/mariadb_43af4698120740069f3c51294b346354/
    Complete output (1 lines):
    MariaDB Connector/Python requires MariaDB Connector/C >= 3.1.3, found version 3.1.13

File: ddtrace/contrib/mariadb/patch.py
Comment (ZStriker19) [reply ddtrace/contrib/mariadb/patch.py:38]: It looks like mariadb's connect function only supports kwargs: https://mariadb-corporation.github.io/mariadb-connector-python/module.html#mariadb.connect

File: tests/contrib/mysql/test_mysql.py
Comment (ZStriker19) [reply tests/contrib/mysql/test_mysql.py]: reverted it, but kept function name change

File: ddtrace/contrib/mariadb/__init__.py
Comment (Kyle-Verhoog) [inline ddtrace/contrib/mariadb/__init__.py:51]: Is this actually true? Seems like it might be copy-paste from the mysql docs.

File: ddtrace/contrib/mariadb/patch.py
Comment (Kyle-Verhoog) [inline ddtrace/contrib/mariadb/patch.py:37]: I don't see where we patch `mariadb.Connect` - we only patch `mariadb.connect`. Should we be patching `Connect` above?

File: releasenotes/notes/mariadb-8e7bc94a0d1b8f02.yaml
Comment (Kyle-Verhoog) [inline releasenotes/notes/mariadb-8e7bc94a0d1b8f02.yaml:4]:
```suggestion
    Add MariaDB integration.

```

Newline

File: tests/contrib/config.py
Comment (Kyle-Verhoog) [inline tests/contrib/config.py:45]: Should we use mariadb specific variables? `TEST_MARIADB_PORT`, etc?

File: tests/contrib/mariadb/test_mariadb.py
Comment (Kyle-Verhoog) [inline tests/contrib/mariadb/test_mariadb.py:23]:
```suggestion
```

File: tests/contrib/mariadb/test_mariadb.py
Comment (Kyle-Verhoog) [inline tests/contrib/mariadb/test_mariadb.py:31]:
```suggestion
```

I think the existence/usage of this function is proof of this fact

File: tests/contrib/mariadb/test_mariadb.py
Comment (Kyle-Verhoog) [inline tests/contrib/mariadb/test_mariadb.py:40]: :+1:

File: tests/contrib/mariadb/test_mariadb.py
Comment (Kyle-Verhoog) [inline tests/contrib/mariadb/test_mariadb.py:226]:
```suggestion
    span = spans[-1]
```

File: tests/contrib/mariadb/test_mariadb.py
Comment (Kyle-Verhoog) [reply tests/contrib/mariadb/test_mariadb.py:253]: I think I actually agree here. We end up testing a large number of properties about the spans in each of these cases and the snapshots are worth it in that case.

File: tests/contrib/mariadb/test_mariadb.py
Comment (Kyle-Verhoog) [inline tests/contrib/mariadb/test_mariadb.py:310]: :+1: this workaround is required since the `snapshot` decorator isn't compatible with `pytest.mark.parametrize` right now unfortunately.

@Yun-Kim not sure if you've happened to dig into some of the pytest internals yet but if so do you think it'd be much work to make `snapshot` compatible with parametrize?

File: tests/contrib/mysql/test_mysql.py
Comment (Kyle-Verhoog) [inline tests/contrib/mysql/test_mysql.py:61]: Let's actually do this change in a separate PR. It seems silly to have to go through all that effort but isolating the changes makes it easier to rollback or cherry-pick in the event that we need to. (ex. if we end up changing something in mysql then we don't want to get merge conflicts because of this mariadb PR)

File: tests/contrib/config.py
Comment (Kyle-Verhoog) [inline tests/contrib/config.py:40]: I can't comment on the file directly (thanks github) but can we delete the `tests/.DS_Store` introduced above?

File: tests/snapshots/tests.contrib.mariadb.test_mariadb.test_query_proc_snapshot.snap
Comment (Kyle-Verhoog) [inline tests/snapshots/tests.contrib.mariadb.test_mariadb.test_query_proc_snapshot.snap:3]: Interesting that the resource for a procedure ends up getting truncated to this... @brettlangdon any thoughts?

File: tests/utils.py
Comment (Kyle-Verhoog) [inline tests/utils.py:883]: :+1: these changes were merged upstream with #2609

**Review by Kyle-Verhoog — CHANGES_REQUESTED** [review]

File: tests/contrib/mysql/test_mysql.py
Comment (ZStriker19) [reply tests/contrib/mysql/test_mysql.py:61]: Makes sense!

File: ddtrace/contrib/mariadb/__init__.py
Comment (ZStriker19) [reply ddtrace/contrib/mariadb/__init__.py:51]: Yeah it is copy and pasted but I though that since it was true for mysql it would also be true for Mariadb.

Comment (ZStriker19) [reply ddtrace/contrib/mariadb/__init__.py:51]: Removing because not relevant

File: tests/contrib/config.py
Comment (ZStriker19) [reply tests/contrib/config.py:40]: removed it

File: ddtrace/contrib/mariadb/__init__.py
Comment (ZStriker19) [reply ddtrace/contrib/mariadb/__init__.py:37]: discussed and changed to saying we need to patch before importing

File: ddtrace/contrib/mariadb/patch.py
Comment (ZStriker19) [reply ddtrace/contrib/mariadb/patch.py:37]: Good catch, Connect is not an exposed function

File: tests/contrib/config.py
Comment (ZStriker19) [reply tests/contrib/config.py:45]: sounds good!

File: tests/contrib/mariadb/test_mariadb.py
Comment (ZStriker19) [reply tests/contrib/mariadb/test_mariadb.py:251]: changed most of them over, left 4 normal tests

File: tests/contrib/mariadb/test_mariadb.py
Comment (Yun-Kim) [reply tests/contrib/mariadb/test_mariadb.py:228]: Sorry, just saw this comment. I haven't dove into `snapshot` with `parameterize` yet, but I'll look into it and let you know if there's an easy compatibility fix.

**Review by Yun-Kim — COMMENTED** [review]
Looks great! Just a couple small questions

File: tests/contrib/mariadb/test_mariadb.py
Comment (Kyle-Verhoog) [reply tests/contrib/mariadb/test_mariadb.py:228]: No rush or anything - just something to think about! We don't need to block this PR on it but we probably want to have them be compatible before we see this pattern 1000 times in the project 😆

File: tests/contrib/mariadb/test_mariadb.py
Comment (Kyle-Verhoog) [inline tests/contrib/mariadb/test_mariadb.py:67]:
```suggestion
def test_query_executemany(connection, tracer):
```

File: tests/contrib/mariadb/test_mariadb.py
Comment (Kyle-Verhoog) [inline tests/contrib/mariadb/test_mariadb.py:68]:
```suggestion
```

File: tests/utils.py
Comment (Kyle-Verhoog) [inline tests/utils.py:920]:
```suggestion
```

**Review by Kyle-Verhoog — CHANGES_REQUESTED** [review]
We forgot to add it, but we need to add an entry on the index page listing the version support for mariadb: https://github.com/DataDog/dd-trace-py/blame/master/docs/index.rst#L94

LGTM once these things have been addressed! :+1::+1:

File: tests/contrib/mariadb/test_mariadb.py
Comment (ZStriker19) [reply tests/contrib/mariadb/test_mariadb.py:97]: It's testing that we're not double tagging with `mariadb.query` and `mariadb.query.fetchall` as keys with the value being the actual query.

File: .gitignore
Comment (Kyle-Verhoog) [inline .gitignore:110]:
```suggestion
```

File: .gitignore
Comment (Kyle-Verhoog) [inline .gitignore:109]:
```suggestion
```

File: tests/contrib/mariadb/test_mariadb.py
Comment (Yun-Kim) [reply tests/contrib/mariadb/test_mariadb.py:97]: Could you point me to the code that sets query span tags? Is it in `dbapi.__init__.py`? I can't seem to find it 😢

Comment (Kyle-Verhoog) [reply tests/contrib/mariadb/test_mariadb.py:97]: I think it's because we're generating two spans with `fetchall` spans and we don't want to double tag the query by tagging the fetchall span.

**Review by Kyle-Verhoog — COMMENTED** [review]
LGTM - there just seems to be a typo or something in the circleci config and then we'll also need #2584 to be merged before CI will pass!

File: Dockerfile.buster
Comment (majorgreys) [inline Dockerfile.buster:59]: Do we need both of these dependencies to run the client?

Comment (ZStriker19) [reply Dockerfile.buster:59]: We may not need `&& curl -sS https://downloads.mariadb.com/MariaDB/mariadb_repo_setup | bash \`  but I'm removing all of these now because the work for the Dockerfile is being done here  #2584

File: tests/contrib/mariadb/test_mariadb.py
Comment (brettlangdon) [inline tests/contrib/mariadb/test_mariadb.py:197]: do these connections need a `with connection: pass`? are they going to get cleaned up after each test case?

File: tests/snapshots/tests.contrib.mariadb.test_mariadb.test_query_proc_snapshot.snap
Comment (brettlangdon) [reply tests/snapshots/tests.contrib.mariadb.test_mariadb.test_query_proc_snapshot.snap:3]: this is probably why: https://github.com/DataDog/dd-trace-py/blob/0dc356f0a672f61383766066de068115d2962fb7/ddtrace/contrib/dbapi/__init__.py#L115

because it is `cursor.callproc("proc_name")` we don't have the query, that is stored on the server

I am surprised we only have 1 span here since this snapshot tests has two executes and the callproc ?

File: tests/contrib/mariadb/test_mariadb.py
Comment (Kyle-Verhoog) [reply tests/contrib/mariadb/test_mariadb.py:42]: Yeah - just thinking that since the test agent isn't 100% yet that it's good to have some plain old python tests around

File: ddtrace/contrib/mariadb/__init__.py
Comment (Kyle-Verhoog) [inline ddtrace/contrib/mariadb/__init__.py:54]:
```suggestion
```

File: ddtrace/contrib/mariadb/__init__.py
Comment (Kyle-Verhoog) [inline ddtrace/contrib/mariadb/__init__.py:3]:
```suggestion
`MariaDB library<https://mariadb-corporation.github.io/mariadb-connector-python/usage.html>`_ to trace queries.
```

File: tests/snapshots/tests.contrib.mariadb.test_mariadb.test_query_proc_snapshot.snap
Comment (brettlangdon) [reply tests/snapshots/tests.contrib.mariadb.test_mariadb.test_query_proc_snapshot.snap:3]: @Kyle-Verhoog any further thoughts on this?

**Review by brettlangdon — APPROVED** [review]
previous comments I'd like to see resolved, otherwise lgtm

**Review by majorgreys — CHANGES_REQUESTED** [review]
small nit

File: tests/contrib/mariadb/test_mariadb.py
Comment (brettlangdon) [inline tests/contrib/mariadb/test_mariadb.py:179]: you could just use your `connection` fixture right? should do the same thing

File: ddtrace/contrib/mariadb/__init__.py
Comment (Kyle-Verhoog) [inline ddtrace/contrib/mariadb/__init__.py:3]:
```suggestion
`MariaDB library <https://mariadb-corporation.github.io/mariadb-connector-python/usage.html>`_ to trace queries.
```

File: tests/snapshots/tests.contrib.mariadb.test_mariadb.test_query_proc_snapshot.snap
Comment (Kyle-Verhoog) [reply tests/snapshots/tests.contrib.mariadb.test_mariadb.test_query_proc_snapshot.snap:3]: hmm yeah haven't thought about it any further but it's definitely non-blocking for this PR

Comment (Kyle-Verhoog) [reply tests/snapshots/tests.contrib.mariadb.test_mariadb.test_query_proc_snapshot.snap:3]: ah for some reason i didn't realize that `sp_sum` was the given name of the stored procedure.. seems totally fine to me :+1:

> I am surprised we only have 1 span here since this snapshot tests has two executes and the callproc ?

lol yeah the test for some reason disables the tracer for the two executes: https://github.com/DataDog/dd-trace-py/pull/2570/files#diff-b5666b23f77bb419456af9d9b0a778a1a787fafbee39303612e5de24e4960500R264

File: Dockerfile.buster
Comment (Kyle-Verhoog) [reply Dockerfile.buster:61]: @ZStriker19 looks like you forgot to push up the revert on this file

**Review by Kyle-Verhoog — COMMENTED** [review]
Literally just one thing left and we're good to go!

**Review by Kyle-Verhoog — APPROVED** [review]
🙌 🙌 🙌

**Review by brettlangdon — APPROVED** [review]

**Pull request merged**
